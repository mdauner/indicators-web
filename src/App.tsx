import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { Indicator, DataSet, IndicatorChoice } from './types';
import Table from './Table';
import { CellProps, Column } from 'react-table';

const INDICATOR_CHOICES: IndicatorChoice[] = [
  {
    id: 'SP.POP.TOTL',
    label: 'Population, total'
  },
  {
    id: 'NY.GDP.MKTP.CD',
    label: 'GDP, total'
  },
  {
    id: 'EN.ATM.CO2E.PC',
    label: 'CO2 Emissions'
  },
  {
    id: 'SP.DYN.LE00.IN',
    label: 'Life Expectancy at Birth'
  },
  {
    id: 'TX.VAL.TECH.MF.ZS',
    label: 'High-technology exports (% of manufactured exports)'
  },
  {
    id: 'IP.PAT.NRES',
    label: 'Patent Application, non-residents'
  },
  {
    id: 'IP.PAT.RESD',
    label: 'Patent Application, residents'
  }
];

function App() {
  const [data, setData] = useState<DataSet[]>();
  const [indicator, setIndicator] = useState<Indicator>('SP.POP.TOTL');
  const [selectedDataSet, selectDataSet] = useState<DataSet>();
  const [selectedYear, selectYear] = useState<string>();

  useEffect(() => {
    fetchData(indicator);
  }, [indicator]);

  const fetchData = async function(indicator: Indicator) {
    let url = `${process.env.REACT_APP_API_BASE_URL}display_data/?indicator=${indicator}`;
    const response = await axios.get<DataSet[]>(url);
    setData(response.data);
  };

  const editCell = React.useCallback(function(dataSet: DataSet, year: string) {
    selectDataSet(dataSet);
    selectYear(year);
  }, []);

  const columns: Column<DataSet>[] = React.useMemo(
    () => [
      {
        Header: 'Country',
        accessor: 'country',
        sticky: 'left',
        Cell: ({ cell }: CellProps<DataSet>) => (
          <div className="font-bold">{cell.value}</div>
        )
      },
      ...Object.keys(data?.[0].data ?? {}).map((year) => ({
        id: year,
        Header: () => <div>{year}</div>,
        accessor: `data[${year}]`,
        Cell: ({ cell }: CellProps<DataSet>) => (
          <div>
            {cell.value
              ? new Intl.NumberFormat('en-US', {
                  minimumFractionDigits: 1,
                  maximumFractionDigits: 1
                }).format(cell.value)
              : '---'}
          </div>
        )
      }))
    ],
    [data]
  );

  return (
    <div className="w-screen flex justify-center">
      <div className="p-8 container mx-auto flex flex-col bg-white my-16 shadow-lg rounded-lg w-full">
        <div className="flex lg:items-center flex-col lg:flex-row">
          <div className="flex-1 text-lg lg:text-3xl font-bold uppercase mb-4 lg:mb-0">
            World Bank data
          </div>
          <div className="flex items-center">
            <label
              className="w-56 lg:text-right lg:mr-2"
              htmlFor="indicator-select"
            >
              Select an indicator:
            </label>
            <select
              id="indicator-select"
              className="block appearance-none w-full bg-gray-300 border border-gray-300 text-gray-700 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
              onChange={(event) =>
                setIndicator(event.target.value as Indicator)
              }
            >
              {INDICATOR_CHOICES.map(({ label, id }) => (
                <option key={id} value={id}>
                  {label}
                </option>
              ))}
            </select>
          </div>
        </div>
        {!data ? (
          <div className="w-full h-64 flex justify-center items-center">
            <div className="font-bold text-2xl">Loading ...</div>
          </div>
        ) : (
          <Table
            columns={columns}
            data={data}
            selectedDataSetId={selectedDataSet?.id}
            selectedYear={selectedYear}
            onClickCell={editCell}
          />
        )}
      </div>
    </div>
  );
}

export default App;
