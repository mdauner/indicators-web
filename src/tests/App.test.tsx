import React from 'react';
import {
  render,
  act,
  waitForElementToBeRemoved,
  fireEvent
} from '@testing-library/react';
import axios from 'axios';

import App, { INDICATOR_CHOICES } from '../App';
import { mockDataSet } from './mocks';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('App', () => {
  beforeEach(() => {
    mockedAxios.get.mockReset();
    mockedAxios.get.mockResolvedValueOnce({ data: [mockDataSet] });
  });

  test('renders loading text initially', async () => {
    await act(async () => {
      const { queryByText } = render(<App />);
      expect(queryByText(/loading/i)).toBeTruthy();
    });
  });

  test('renders data', async () => {
    await act(async () => {
      const { container, queryByText } = render(<App />);
      await waitForElementToBeRemoved(() => queryByText(/loading/i));

      expect(container.firstChild).toMatchSnapshot();
    });
    expect(mockedAxios.get).toHaveBeenCalled();
  });

  test('loads new data when indicator was selected', async () => {
    let indicatorSelect: HTMLElement;
    await act(async () => {
      const { getByLabelText, queryByText } = render(<App />);
      await waitForElementToBeRemoved(() => queryByText(/loading/i));

      indicatorSelect = getByLabelText(/indicator/i);
    });
    await act(async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: [mockDataSet] });

      fireEvent.change(indicatorSelect, {
        target: { value: INDICATOR_CHOICES[2].id }
      });
    });
    expect(mockedAxios.get).toHaveBeenCalledWith(
      expect.stringContaining(INDICATOR_CHOICES[2].id)
    );
  });

  test('shows editor when cell is selected', async () => {
    await act(async () => {
      const { queryByTestId, getByText, queryByText } = render(<App />);
      await waitForElementToBeRemoved(() => queryByText(/loading/i));

      expect(queryByTestId('editor')).not.toBeInTheDocument();

      const cell = getByText(mockDataSet.data['1960']);
      fireEvent.click(cell);

      expect(queryByTestId('editor')).toBeInTheDocument();
    });
  });

  test('sorts table when header is clicked', async () => {
    await act(async () => {
      const { container, getByText, queryByText } = render(<App />);
      await waitForElementToBeRemoved(() => queryByText(/loading/i));

      const header = getByText('1960');
      fireEvent.click(header);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
