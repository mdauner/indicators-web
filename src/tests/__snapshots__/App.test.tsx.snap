// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders data 1`] = `
<div
  class="w-screen flex justify-center"
>
  <div
    class="p-8 container mx-auto flex flex-col bg-white my-16 shadow-lg rounded-lg w-full"
  >
    <div
      class="flex lg:items-center flex-col lg:flex-row"
    >
      <div
        class="flex-1 text-lg lg:text-3xl font-bold uppercase mb-4 lg:mb-0"
      >
        World Bank data
      </div>
      <div
        class="flex items-center"
      >
        <label
          class="w-56 lg:text-right lg:mr-2"
          for="indicator-select"
        >
          Select an indicator:
        </label>
        <select
          class="block appearance-none w-full bg-gray-300 border border-gray-300 text-gray-700 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
          id="indicator-select"
        >
          <option
            value="SP.POP.TOTL"
          >
            Population, total
          </option>
          <option
            value="NY.GDP.MKTP.CD"
          >
            GDP, total
          </option>
          <option
            value="EN.ATM.CO2E.PC"
          >
            CO2 Emissions
          </option>
          <option
            value="SP.DYN.LE00.IN"
          >
            Life Expectancy at Birth
          </option>
          <option
            value="TX.VAL.TECH.MF.ZS"
          >
            High-technology exports (% of manufactured exports)
          </option>
          <option
            value="IP.PAT.NRES"
          >
            Patent Application, non-residents
          </option>
          <option
            value="IP.PAT.RESD"
          >
            Patent Application, residents
          </option>
        </select>
      </div>
    </div>
    <div
      class="overflow-x-scroll"
    >
      <table
        class="table sticky text-left w-full border-collapse border mt-4 p-8"
        role="table"
      >
        <thead
          class="header"
        >
          <tr
            role="row"
          >
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              data-sticky-last-left-td="true"
              role="columnheader"
              style="position: sticky; z-index: 2; left: 0px; cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                Country
                <div
                  class="ml-2"
                />
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1960
                </div>
                <div
                  class="ml-2"
                />
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1961
                </div>
                <div
                  class="ml-2"
                />
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1962
                </div>
                <div
                  class="ml-2"
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="body"
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  "
              data-sticky-last-left-td="true"
              role="cell"
              style="position: sticky; z-index: 2; left: 0px;"
            >
              <div
                class="font-bold"
              >
                USA
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                123.3
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                125.2
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                131.3
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`App sorts table when header is clicked 1`] = `
<div
  class="w-screen flex justify-center"
>
  <div
    class="p-8 container mx-auto flex flex-col bg-white my-16 shadow-lg rounded-lg w-full"
  >
    <div
      class="flex lg:items-center flex-col lg:flex-row"
    >
      <div
        class="flex-1 text-lg lg:text-3xl font-bold uppercase mb-4 lg:mb-0"
      >
        World Bank data
      </div>
      <div
        class="flex items-center"
      >
        <label
          class="w-56 lg:text-right lg:mr-2"
          for="indicator-select"
        >
          Select an indicator:
        </label>
        <select
          class="block appearance-none w-full bg-gray-300 border border-gray-300 text-gray-700 py-3 px-4 pr-8 rounded leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
          id="indicator-select"
        >
          <option
            value="SP.POP.TOTL"
          >
            Population, total
          </option>
          <option
            value="NY.GDP.MKTP.CD"
          >
            GDP, total
          </option>
          <option
            value="EN.ATM.CO2E.PC"
          >
            CO2 Emissions
          </option>
          <option
            value="SP.DYN.LE00.IN"
          >
            Life Expectancy at Birth
          </option>
          <option
            value="TX.VAL.TECH.MF.ZS"
          >
            High-technology exports (% of manufactured exports)
          </option>
          <option
            value="IP.PAT.NRES"
          >
            Patent Application, non-residents
          </option>
          <option
            value="IP.PAT.RESD"
          >
            Patent Application, residents
          </option>
        </select>
      </div>
    </div>
    <div
      class="overflow-x-scroll"
    >
      <table
        class="table sticky text-left w-full border-collapse border mt-4 p-8"
        role="table"
      >
        <thead
          class="header"
        >
          <tr
            role="row"
          >
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              data-sticky-last-left-td="true"
              role="columnheader"
              style="position: sticky; z-index: 2; left: 0px; cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                Country
                <div
                  class="ml-2"
                />
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1960
                </div>
                <div
                  class="ml-2"
                >
                   ðŸ”¼
                </div>
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1961
                </div>
                <div
                  class="ml-2"
                />
              </div>
            </th>
            <th
              class="bg-white py-4 px-6 bg-grey-lightest font-bold uppercase text-sm text-grey-dark border-b border-r border-gray-300"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              <div
                class="flex items-center"
              >
                <div>
                  1962
                </div>
                <div
                  class="ml-2"
                />
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="body"
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  "
              data-sticky-last-left-td="true"
              role="cell"
              style="position: sticky; z-index: 2; left: 0px;"
            >
              <div
                class="font-bold"
              >
                USA
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                123.3
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                125.2
              </div>
            </td>
            <td
              class="py-4 px-6 border-b border-r border-gray-300 text-xs bg-white text-right  cursor-pointer "
              role="cell"
            >
              <div>
                131.3
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
